beartype/typing.py

```diff
@@ -288,3 +280,197 @@
+        Stand-in for :class:`typing.Protocol`'s metaclass that caches results of
+        :meth:`class.__instancecheck__`, (which is otherwise `really expensive
+        <https://github.com/python/mypy/issues/3186#issuecomment-885718629>`.
+        The downside is that this will yield unpredictable results for objects
+        whose methods don't stem from any type (e.g., are assembled at runtime).
+        This is ill-suited for such "types".
+
+        Note that one can make an existing protocol a caching protocol through
+        inheritance, but in order to be ``@runtime_checkable``, the parent
+        protocol also has to be ``@runtime_checkable``.
+
+        .. code-block:: python
+          :linenos:
+
+          >>> from abc import abstractmethod
+          >>> from typing import Protocol
+          >>> from beartype.typing import _CachingProtocolMeta, runtime_checkable
+
+          >>> @runtime_checkable
+          ... class _MyProtocol(Protocol):  # plain vanilla protocol
+          ...   @abstractmethod
+          ...   def myfunc(self, arg: int) -> str:
+          ...     pass
+
+          >>> @runtime_checkable  # redundant, but useful for documentation
+          ... class MyProtocol(
+          ...   _MyProtocol,
+          ...   Protocol,
+          ...   metaclass=_CachingProtocolMeta,  # caching version
+          ... ):
+          ...   pass
+
+          >>> class MyImplementation:
+          ...   def myfunc(self, arg: int) -> str:
+          ...     return str(arg * -2 + 5)
+
+          >>> my_thing: MyProtocol = MyImplementation()
+          >>> isinstance(my_thing, MyProtocol)
+          True
+
+        The easy way to ensure your protocol caches checks and is
+         ``@runtime_checkable`` is to inherit from
+         :class:`beartype.typing.Protocol` instead:
+
+        .. code-block:: python
+          :linenos:
+
+          >>> from beartype.typing import Protocol
+
+          >>> class MyBearProtocol(Protocol):
+          ...   @abstractmethod
+          ...   def myfunc(self, arg: int) -> str:
+          ...     pass
+
+          >>> my_thing: MyBearProtocol = MyImplementation()
+          >>> isinstance(my_thing, MyBearProtocol)
+          True
+        """
+
+        _abc_inst_check_cache: Dict[Type, bool]
+
+        def __new__(
+            mcls: Type[_TT],
+            name: str,
+            bases: Tuple[Type, ...],
+            namespace: Dict[str, Any],
+            **kw: Any,
+        ) -> _TT:
+            # See <https://github.com/python/mypy/issues/9282>
+            cls = super().__new__(mcls, name, bases, namespace, **kw)  # type: ignore [misc]
+            # Prefixing this class member with "_abc_" is necessary to prevent
+            # it from being considered part of the Protocol. (See
+            # <https://github.com/python/cpython/blob/main/Lib/typing.py>.)
+            cls._abc_inst_check_cache = {}
+
+            return cls
+
+        def __instancecheck__(cls, inst: Any) -> bool:
```

Comment:
This implementation is ported straight across from `numerary` and does not limit entries, which we'll likely want to introduce as part of this PR. I thought treating this as a functioning placeholder was a good place to start, though.